### PR TITLE
[SPARK-8498][TUNGSTEN] fix npe in errorhandling path in unsafeshuffle writer

### PR DESCRIPTION
**Pull request by holdenk**

### CODE REVIEW

File: core/src/test/java/org/apache/spark/shuffle/unsafe/UnsafeShuffleWriterSuite.java [hunk 1]
```diff
@@ -253,6 +253,23 @@ public void doNotNeedToCallWriteBeforeUnsuccessfulStop() throws IOException {
     createWriter(false).stop(false);
   }
 
+  class PandaException extends RuntimeException {
+  }
+
+  @Test(expected=PandaException.class)
+  public void writeFailurePropegates() throws Exception {
```
Comment (JoshRosen) [inline core/src/test/java/org/apache/spark/shuffle/unsafe/UnsafeShuffleWriterSuite.java]: Spelling: I think this should be "propagates".

File: core/src/main/java/org/apache/spark/shuffle/unsafe/UnsafeShuffleWriter.java [hunk 2]
```diff
@@ -139,17 +139,22 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
...
     try {
       while (records.hasNext()) {
         insertRecordIntoSorter(records.next());
       }
       closeAndWriteOutput();
-      success = true;
-    } finally {
-      if (!success) {
-        sorter.cleanupAfterError();
+    } catch (Exception e) {
```
Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/shuffle/unsafe/UnsafeShuffleWriter.java]: AFAIK my original intent was to also ensure that we clean up after Throwables as well without necessarily having to catch throwable, so I think that's why I had the slightly unclear `success` boolean.

**Comment by JoshRosen** [comment]
Thanks for fixing this; the test looks good to me, but I think we might want to preserve the old behavior of also performing error cleanup when Throwables are thrown (i.e. by using the `success` flag idiom that I used before, but perhaps with a comment this time to explain what's going on).

**Comment by holdenk** [comment]
Sounds like a good plan.

**Comment by holdenk** [comment]
Would it not make more sense to just catch/rethrow throwables? (That way if fail during cleanup we can pass back an exception indicating that cleanup failed + the underlying cause as well)?

**Comment by JoshRosen** [comment]
I think that my concern was that wrapping a Throwable with a cause and re-throwing it might not be safe in general because catchers higher up the stack would then receive a different exception than the original Throwable, which might lead to problems if the throwable was an Error that's treated specially (e.g. ThreadDeath).

I suppose one approach would be to not re-wrap the Throwable when re-throwing. but AFAIK the compiler will complain if you re-throw `Throwable` in a method that doesn't declare that it `throws Throwable`. I think there's a reference for this in Guava's Throwables documentation: https://code.google.com/p/guava-libraries/wiki/ThrowablesExplained.  It looks like this is a Java 6-ism, though, since Java 7 seems to allow this type of re-throwing: https://stackoverflow.com/questions/7428966/how-does-the-correct-rethrow-code-look-like-for-this-example

It looks like Java 7 provides us with a nice way to handle this with try-with-resources (see https://stuartmarks.wordpress.com/2011/07/21/a-new-java-exception-handling-idiom/), but I don't think we can use this here as long as we want to backport this fix to 1.4.1, which still must build for Java 6.

I think that's why I originally decided to go with the `finally` block, although I guess I should have added more try blocks to guard against errors during the cleanup itself.

File: core/src/main/java/org/apache/spark/shuffle/unsafe/UnsafeShuffleWriter.java [hunk 3]
```diff
@@ -147,8 +150,16 @@ public void write(scala.collection.Iterator<Product2<K, V>> records) throws IOEx
       closeAndWriteOutput();
       success = true;
     } finally {
-      if (!success) {
-        sorter.cleanupAfterError();
+      if (sorter != null) {
+        try {
+          sorter.cleanupAfterError();
+        } catch (Exception e) {
```
Comment (stephenh) [inline core/src/main/java/org/apache/spark/shuffle/unsafe/UnsafeShuffleWriter.java]: You could potentially use Utils.tryWithSafeFinally, which does this sort of annoying "try/catch in a try/catch" thing.

Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/shuffle/unsafe/UnsafeShuffleWriter.java]: This class, along with a lot of the other Tungsten code, is written in Java, so we can't use `Utils.tryWithSafeFinally` here.

File: core/src/main/java/org/apache/spark/shuffle/unsafe/UnsafeShuffleWriter.java [hunk 4]
```diff
@@ -147,8 +150,16 @@ public void write(scala.collection.Iterator<Product2<K, V>> records) throws IOEx
...
+          // Only throw this error if we won't be masking another
+          // error.
+          if (success) {
+            throw e;
+          }
```
Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/shuffle/unsafe/UnsafeShuffleWriter.java]: Shouldn't we log in the `else` branch to make sure that the error is never silently swallowed?

**Comment by JoshRosen** [comment]
LGTM pending Jenkins; thanks!

**Comment by JoshRosen** [comment]
Merging this to `master` and `branch-1.4`.

**Comment by JoshRosen** [comment]
Dammit, I just realized that the `catch Exception` added in the latest commit will break compilation under Java 6 in branch-1.4 :(.  I'm going to revert in branch-1.4 and will just push a hotfix later to address this (using either Guava or Unsafe to handle re-throw).

**Pull request closed without merging**